### PR TITLE
Add an option to force the numeric type of a field sort

### PR DESCRIPTION
**Pull request by jimczi**
This change adds an option to the `FieldSortBuilder` that allows to transform the type
of a numeric field into another. Possible values for this option are `long` that transforms
the source field into an integer and `double` that transforms the source field into a floating point.
This new option is useful for cross-index search when the sort field is mapped differently on some
indices. For instance if a field is mapped as a floating point in one index and as an integer in another
it is possible to align the type for both indices using the `numeric_type` option:

```
{
   "sort": {
    "field": "my_field",
    "numeric_type": "double" <1>
   }
}
```

<1> Ensure that values for this field are transformed to a floating point if needed.

Only `long` and `double` are supported at the moment but the goal is to also handle `date` and `date_nanos` when #32601 is merged.

Naming should also be improved, I choose `numeric_type` to start with something but I am not happy with it. Propositions are welcome ;).

### CODE REVIEW

**Review by mayya-sharipova — APPROVED** [review]
Thanks Jim, makes sense!

One thing I was wondering (not related to your PR) if it is correct that we return 503 for wrong sorting (using float and double and not setting `numeric_type`)?  Should it be 400 instead?

```js
{
  "error": {
    "root_cause": [],
    "type": "search_phase_execution_exception",
    "reason": "",
    "phase": "fetch",
    "grouped": true,
    "failed_shards": [],
    "caused_by": {
      "type": "class_cast_exception",
      "reason": "class java.lang.Float cannot be cast to class java.lang.Double (java.lang.Float and java.lang.Double are in module java.base of loader 'bootstrap')"
    }
  },
  "status": 503
}
```

File: docs/reference/search/request/sort.asciidoc [hunk 1]
```diff
@@ -108,6 +108,73 @@ POST /_search
...
+// CONSOLE
+// TEST[continued]
+
+In the example above, values for the `index_long` index are casted to
+a double in order to be compatible with the values produced by the
+`index_double` index.
+It is also possible to transform a floating point field into a `long`
+but note that in this case floating points are replaced by the largest
+value that is less than or equal to the argument and is equal to a mathematical
+integer.
```
Comment (jpountz) [inline docs/reference/search/request/sort.asciidoc]: this is not correct for negative values

File: server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java [hunk 2]
```diff
@@ -351,7 +397,18 @@ public SortFieldAndFormat build(QueryShardContext context) throws IOException {
                     && (sortMode == SortMode.SUM || sortMode == SortMode.AVG || sortMode == SortMode.MEDIAN)) {
                 throw new QueryShardException(context, "we only support AVG, MEDIAN and SUM on number based fields");
             }
-            SortField field = fieldData.sortField(missing, localSortMode, nested, reverse);
+            final SortField field;
+            if (numericType != null) {
+                if (fieldData instanceof SortedNumericDVIndexFieldData == false) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java]: why do you check against this class rather than IndexNumericFieldData?

**Review by jpountz — CHANGES_REQUESTED** [review]
I left some comments but the idea makes sense to me.

File: docs/reference/search/request/sort.asciidoc (hunk 1 above)
Comment (jimczi) [reply docs/reference/search/request/sort.asciidoc]: right, thanks

File: server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java]: This is a mistake, I pushed https://github.com/elastic/elasticsearch/pull/38095/commits/24c961afd1d7b6e0197d8d15e015761ffeae6d7d

**Review by jpountz — APPROVED** [review]

**Comment by jpountz** [comment]
@jimczi  I'm assuming there is nothing left to backport and removed the backport pending label.

**Comment by russcam** [comment]
Looks like this went into v7.2.0? https://github.com/elastic/elasticsearch/blob/be513ce15956f552792fdfe0d661ee65b163dd67/server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java#L306-L321

**Comment by jpountz** [comment]
@russcam Indeed, I just fixed the version tag and the release notes. Thanks for spotting!

**Pull request merged**